experiments/optimizer/run.py

```diff
@@ -0,0 +1,413 @@
+  
+def make_model(pretrained_name, max_tokens, tokenizer , use_flash):
+
+    config = AutoConfig.from_pretrained(
+        pretrained_name,
+        vocab_size=len(tokenizer),
+        n_ctx=max_tokens,
+        bos_token_id=tokenizer.bos_token_id,
+        eos_token_id=tokenizer.eos_token_id,
+        pad_token_id=tokenizer.pad_token_id,
+       optimize_cuda_cache=True
+    )
+    #model = GPT2LMHeadModel.from_pretrained(pretrained_name).cuda()
+    model = GPT2LMHeadModel(config).cuda()
+    GPT2Attention._attn = _attn_orig
+    if use_flash:
+        print("Use Flash Attention")
+        GPT2Attention._attn = _attn_wrapper
+
+    # model.config.update(
+    #     dict(
+    #         n_ctx=max_tokens,
+    #         #n_positions=max_tokens,
+    #         vocab_size=len(tokenizer),
+    #         bos_token_id=tokenizer.bos_token_id,
+    #          eos_token_id=tokenizer.eos_token_id,
+    #           pad_token_id=tokenizer.pad_token_id,
+    #          optimize_cuda_cache=True
+    #     )
+    # )
+    model.resize_token_embeddings(len(tokenizer))
+
+    model_size = sum(t.numel() for t in model.parameters())
+    print(f"GPT-2 size: {model_size/1000**2:.1f}M parameters")
+
+    model_size = sum(p.numel() for p in model.parameters() if p.requires_grad)
+    print(f"GPT-2 size requires_grad: {model_size/1000**2:.1f}M parameters")
+    # # patch model embeddings
+    # emb = model.transformer.wpe.weight.data
+    # wpe = nn.Embedding(max_tokens, emb.shape[1])
+    # wpe.weight.data = emb.repeat(max_tokens // emb.shape[0], 1)
+    # model.transformer.wpe = wpe
+
+    # # also increase mask size
+    # for block in model.transformer.h:
+    #     block.attn.bias.data = (
+    #         torch.tril(torch.ones((max_tokens, max_tokens), dtype=torch.bool))
+    #         .view(1, 1, max_tokens, max_tokens)
+    #         .cuda()
+    #     )
+    if torch.cuda.device_count() > 1:
+            print("Let's use", torch.cuda.device_count(), "GPUs!")
+            model = nn.DataParallel(model)
+    return model
+
+
+class DatasetWrapper(IterableDataset):
+    def __init__(self, max_tokens=256):
+        self.tokenizer = GPT2TokenizerFast.from_pretrained("flax-community/gpt2-base-thai" , bos_token='<|startoftext|>', eos_token='<|endoftext|>', pad_token='<|pad|>')
+        self.max_tokens = max_tokens
+
+    def __iter__(self):
+        buffer = []
+        for sample in load_dataset("mc4",
+                       languages=["th"], 
+                        streaming=True, # optional
+                        split="train").shuffle(buffer_size=10_000):
+            buffer += self.tokenizer( sample["text"] + '<|endoftext|>')["input_ids"]
+            #buffer += [self.tokenizer.eos_token_id]
+            #print(self.tokenizer.decode(buffer))
+            while len(buffer) > self.max_tokens:
+                yield torch.tensor(buffer[: self.max_tokens])
+                buffer = buffer[self.max_tokens :]
+
+class DatasetWrapper_val(IterableDataset):
```

Comment:
How different between `DatasetWrapper` and `DatasetWrapper_val`? Could we refactor this?

Reply:
Done